### PR TITLE
[2/2] Implement macro meta-variable expression

### PR DESCRIPTION
**Pull request by c410-f3r**
Final part of https://github.com/rust-lang/rust/pull/93545#issuecomment-1050963295

r? @petrochenkov 

cc #83527

### CODE REVIEW

File: compiler/rustc_expand/src/mbe/transcribe.rs [hunk 1]
```diff
@@ -411,13 +412,152 @@ fn lockstep_iter_size(
...
+    for &(idx, _) in repeats {
+        match matched {
+            MatchedSeq(ref ads) if let Some(elem) = ads.get(idx) => matched = elem,
+           // We can only get here if `repeats` is more deeply nested than the
+           // `NamedMatch` passed in to `count_repetitions`. That would
+           // indicate that the RHS of the macro is more deeply nested than
+           // the LHS, which is an error.
+            _ => return Err(cx.struct_span_err(
+                sp.entire(),
+                &format!("variable `{}` does not repeat at this depth", ident),
```
Comment (c410-f3r) [inline compiler/rustc_expand/src/mbe/transcribe.rs:485]: Couldn't reproduce after some attempts but I will try again tomorrow

File: compiler/rustc_expand/src/mbe/transcribe.rs [hunk 2]
```diff
@@ -411,13 +412,152 @@ fn lockstep_iter_size(
...
+                            sp.entire(),
+                        )
+                        .into(),
+                    );
+                }
+                None => {
+                    return Err(cx.struct_span_err(
+                        sp.entire(),
+                        &format!(
+                            "variable `{}` is not recognized in meta-variable expression",
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply compiler/rustc_expand/src/mbe/transcribe.rs]: This, as well as the above error, can not be triggered due to previous different error checks. As a result, both are now simply unwrapping their desired collections.

I could be wrong here but after several failed attempts, I really couldn't trigger them

File: compiler/rustc_expand/src/mbe/transcribe.rs [hunk 3]
```diff
@@ -411,13 +412,131 @@ fn lockstep_iter_size(
...
+                        "`count` can not be placed inside the inner-most repetition",
+                    ));
+                }
+                match depth_opt {
+                    None => Ok(1),
+                    Some(_) => Err(out_of_bounds_err(cx, system_depth, sp.entire(), "count")),
+                }
+            }
+            MatchedSeq(ref named_matches) => {
+                let new_system_depth = system_depth.wrapping_add(1);
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply compiler/rustc_expand/src/mbe/transcribe.rs:454]: Hum... I have always wondered about the possible effects of debug and release compiler builds in a user's codebase. Well, I guess this and other compiler arithmetic operations will always wrap regardless 🤷

File: compiler/rustc_expand/src/mbe/transcribe.rs [hunk 4]
```diff
@@ -411,13 +412,131 @@ fn lockstep_iter_size(
...
+    sp: &DelimSpan,
+) -> PResult<'a, usize> {
+    // Recursively count the number of matches in `matched` at given depth
+    // (or at the top-level of `matched` if no depth is given).
+    fn count<'a>(
+        cx: &ExtCtxt<'a>,
+        depth_opt: Option<usize>,
+        matched: &NamedMatch,
+        sp: &DelimSpan,
+        system_depth: usize,
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply compiler/rustc_expand/src/mbe/transcribe.rs:438]: The number of declared **nested** loops. Changed to `declared_lhs_depth`

File: compiler/rustc_expand/src/mbe/transcribe.rs [hunk 5]
```diff
@@ -411,13 +412,131 @@ fn lockstep_iter_size(
...
+    interp: &FxHashMap<MacroRulesNormalizedIdent, NamedMatch>,
+    repeats: &[(usize, usize)],
+    result: &mut Vec<TreeAndSpacing>,
+    sp: &DelimSpan,
 ) -> PResult<'a, ()> {
+    match expr {
+        mbe::MetaVarExpr::Count(original_ident, depth_opt) => {
+            let ident = MacroRulesNormalizedIdent::new(original_ident);
+            let count =
+                count_repetitions(cx, depth_opt, interp.get(&ident).unwrap(), &repeats, sp)?;
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply compiler/rustc_expand/src/mbe/transcribe.rs:506]: OMG 🤦

Thanks @petrochenkov

File: compiler/rustc_expand/src/mbe/transcribe.rs [hunk 6]
```diff
@@ -411,13 +412,131 @@ fn lockstep_iter_size(
...
+                    TokenTree::token(
+                        token::TokenKind::lit(token::Integer, sym::integer(*index), None),
+                        sp.entire(),
+                    )
+                    .into(),
+                );
+            }
+            None => return Err(out_of_bounds_err(cx, repeats.len(), sp.entire(), "index")),
+        },
+        mbe::MetaVarExpr::Length(depth) => match repeats.iter().nth_back(depth) {
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply compiler/rustc_expand/src/mbe/transcribe.rs:528]: Removed all but `mbe::Integer`

File: compiler/rustc_expand/src/mbe/transcribe.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (c410-f3r) [reply compiler/rustc_expand/src/mbe/transcribe.rs:506]: OK. Fixed

Also added the same check for `${ignore(SOME_UNKNOWN_IDENT)}`.

File: compiler/rustc_expand/src/mbe/transcribe.rs [hunk 7]
```diff
@@ -411,13 +411,150 @@ fn lockstep_iter_size(
...
-    _sp: &DelimSpan,
+    cx: &ExtCtxt<'a>,
+    expr: MetaVarExpr,
+    interp: &FxHashMap<MacroRulesNormalizedIdent, NamedMatch>,
+    repeats: &[(usize, usize)],
+    result: &mut Vec<TreeAndSpacing>,
+    sp: &DelimSpan,
 ) -> PResult<'a, ()> {
+    match expr {
+        mbe::MetaVarExpr::Count(original_ident, depth_opt) => {
```
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/transcribe.rs:521]:
```suggestion
        MetaVarExpr::Count(original_ident, depth_opt) => {
```

File: compiler/rustc_expand/src/mbe/transcribe.rs [hunk 8]
```diff
@@ -411,13 +411,150 @@ fn lockstep_iter_size(
...
+            let matched = matched_from_ident(cx, original_ident, interp)?;
+            let count = count_repetitions(cx, depth_opt, matched, &repeats, sp)?;
+            let tt = TokenTree::token(
+                token::TokenKind::lit(token::Integer, sym::integer(count), None),
```
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/transcribe.rs:525]:
```suggestion
                TokenKind::lit(token::Integer, sym::integer(count), None),
```

**Comment by petrochenkov** [comment]
r=me with the remaining nits addressed.

File: compiler/rustc_expand/src/mbe/transcribe.rs (hunk 8 above)
Comment (c410-f3r) [reply compiler/rustc_expand/src/mbe/transcribe.rs]: Strange. I could swear they were removed

**Pull request merged**